client/components/data/query/index.jsx

```diff
@@ -0,0 +1,59 @@
+/** @format */
+
+/**
+ * External dependencies
+ */
+import PropTypes from 'prop-types';
+import { PureComponent } from 'react';
+import { bindActionCreators } from 'redux';
+import { connect } from 'react-redux';
+import { values } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import compareProps from 'lib/compare-props';
+import { getRequest } from 'state/selectors';
+
+class Query extends PureComponent {
+	static propTypes = {
+		compareOptions: PropTypes.object,
+		requestAction: PropTypes.func.isRequired,
+		requesting: PropTypes.bool.isRequired,
+	};
+
+	componentWillMount() {
+		this.comparator = compareProps( {
+			ignore: [ 'compareOptions', 'requestAction', 'requesting' ],
+			...this.props.compareOptions,
+		} );
+		this.request( this.props );
+	}
+
+	componentWillReceiveProps( nextProps ) {
+		if ( ! this.comparator( this.props, nextProps ) ) {
+			this.request( nextProps );
+		}
+	}
+
+	// eslint-disable-next-line no-unused-vars
+	request( { compareOptions, requestAction, requesting, ...props } ) {
+		if ( requesting ) {
```

Comment:
I wish we could get rid of the `requesting` prop and rely on the built-in deduplication in data layer.

Deduplication inside the `Query` component is very often the only reason why a `requesting` flag exists in a reducer, and we could avoid a lot of overhead by not implementing it.

After dispatching a request action in a lifecycle method, it takes a surprisingly complicated series of steps before a `requesting` prop of a `Query` component is updated to `true`. And it can lead to subtle bugs like the one fixed in #21858. Maintaining the `requesting` flag outside React/Redux is much simpler.